src/cgroups/v2/systemd_manager.rs

```diff
@@ -0,0 +1,304 @@
+use std::{
+    fs::{self},
+    os::unix::fs::PermissionsExt,
+};
+
+use anyhow::{anyhow, bail, Result};
+use nix::unistd::Pid;
+use oci_spec::LinuxResources;
+use std::path::{Path, PathBuf};
+
+use super::{cpu::Cpu, cpuset::CpuSet, hugetlb::HugeTlb, io::Io, memory::Memory, pids::Pids};
+use crate::cgroups::common;
+use crate::cgroups::common::{write_cgroup_file, CgroupManager};
+use crate::cgroups::v2::controller::Controller;
+use crate::cgroups::v2::controller_type::ControllerType;
+use crate::utils::PathBufExt;
+
+const CGROUP_PROCS: &str = "cgroup.procs";
+const CGROUP_CONTROLLERS: &str = "cgroup.controllers";
+const CGROUP_SUBTREE_CONTROL: &str = "cgroup.subtree_control";
+
+// v2 systemd only supports cpu, io, memory and pids.
+const CONTROLLER_TYPES: &[ControllerType] = &[
+    ControllerType::Cpu,
+    ControllerType::Io,
+    ControllerType::Memory,
+    ControllerType::Pids,
+];
+
+/// SystemDCGroupManager is a driver for managing cgroups via systemd.
+pub struct SystemDCGroupManager {
+    root_path: PathBuf,
+    cgroups_path: CgroupsPath,
+}
+
+/// Represents the systemd cgroups path:
+/// It should be of the form [slice]:[scope_prefix]:[name].
+/// The slice is the "parent" and should be expanded properly,
+/// see expand_slice below.
+struct CgroupsPath {
+    parent: String,
+    scope: String,
+    name: String,
+}
+
+impl SystemDCGroupManager {
+    pub fn new(root_path: PathBuf, cgroups_path: PathBuf) -> Result<Self> {
+        // cgroups path may never be empty as it is defaulted to `/youki`
+        // see 'get_cgroup_path' under utils.rs.
+        // if cgroups_path was provided it should be of the form [slice]:[scope_prefix]:[name],
+        // for example: "system.slice:docker:1234".
+        let mut parent = "";
+        let scope;
+        let name;
+        if cgroups_path.starts_with("/youki") {
+            scope = "youki";
+            name = cgroups_path
+                .strip_prefix("/youki/")?
+                .to_str()
+                .ok_or_else(|| anyhow!("Failed to parse cgroupsPath field."))?;
+        } else {
+            let parts = cgroups_path
+                .to_str()
+                .ok_or_else(|| anyhow!("Failed to parse cgroupsPath field."))?
+                .split(':')
+                .collect::<Vec<&str>>();
+            parent = parts[0];
+            scope = parts[1];
+            name = parts[2];
+        }
+
+        // TODO: create the systemd unit using a dbus client.
+
+        Ok(SystemDCGroupManager {
+            root_path,
+            cgroups_path: CgroupsPath {
+                parent: parent.to_string(),
```

Comment:
Nit: Use to_owned instead of to_string to communicate that you create an owned String.

Reply:
Done.